File: backtrack/77.combinations.cpp
/*
 * @lc app=leetcode id=77 lang=cpp
 *
 * [77] Combinations
 */

#include "algorithm"
#include "iostream"
#include "map"
#include "numeric"
#include "queue"
#include "set"
#include "vector"

using namespace std;

// @lc code=start
class Solution {
  vector<vector<int>> res;

public:
  vector<vector<int>> combine(int n, int k) {
    backtrack(1, n, k, {});
    return res;
  }

  void backtrack(int first, int n, int k, vector<int> comb) {
    if (comb.size() == k) {
      res.push_back(comb);
    }

    for (size_t i = first; i < n + 1; ++i) {
      comb.push_back(i);
      backtrack(i + 1, n, k, comb);
      comb.pop_back();
    }
  }
};
// @lc code=end


int main(){
  cout << "hello world" << endl;
}

File: backtrack/77.combinations.py
#
# @lc app=leetcode id=77 lang=python3
#
# [77] Combinations
#

# @lc code=start
from itertools import combinations
from pprint import pprint
from typing import List


class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        def backtrack(first=1, curr=[]):
            # if combination is done
            if len(curr) ==k:
                output.append(curr[:])
            
            for i in range(first,n+1):
                curr.append(i)
                
                backtrack(i+1,curr)
                curr.pop()
                
                
        output = []
        backtrack()
        return output

# @lc code=end

s = Solution()
pprint(s.combine(4,2))


class Solution2:
    def combine(self, n: int, k: int) -> List[List[int]]:
        return list(combinations(range(1, n+1), k))

